Automattic/wp-calypso
# FSE: Update blocks categories

#### Changes proposed in this Pull Request

- **Update the category names for FSE blocks**. More details: pc6Uok-1b-p2 and [here](https://github.com/Automattic/wp-calypso/issues/43198) (second task listed in the issue). 
-  **Adds a helper function that is used to safely update category names while still supporting old environments**
- **Fix the FSE Jest configuration for TypeScript**. It was needed in order for to properly transpile the test and the helper function included in the changeset when they are consumed by Jest. I decided not to split this change in another PR for two reasons: it's small and it's directly related to the change here.

#### Related PRs
* Jetpack blocks: https://github.com/Automattic/jetpack/pull/16363
* wpcom-blocks: https://github.com/Automattic/wpcom-blocks/pull/178
* block-experiments: https://github.com/Automattic/block-experiments/pull/110

#### Categories

- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-dc263ccf11dec7e5cc2cd31b6672b8dfR35) newspack's `a8c/blog-posts` from `layout` -> `widgets`;
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-05d7ff5647d5a588247b7c0a0bb98bf9R36) newspack's `a8c/posts-carousel` from `layout` -> `widgets`*[0]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-53b5a98f30dc7e6d356d4dae754c02daR24) `a8c/template` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-6893b1bc3b7ff82118bc9334f2e83bbbR26) `a8c/navigation-menu` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-df04a9b006edc7c589a5d70ee6f61e4bR22) `a8c/post-content` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-3b05c15146c5a00d3547626c18b7ffecR19) `a8c/site-title` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-37e1c83f9166b0a7f9aecc29345b4050R22)`a8c/site-credit` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-029c4ccfa1c3a256ee21db826f2cc496R24)  `a8c/site-description` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-d8422954d098352028006c46af21bdffR52) `a8c/donations` from `common` -> `widgets`*[0]
- [X] Keep `jetpack/event-countdown` in the `widgets` category
- [X] Keep `jetpack/timeline` in the `widgets` category
- [X] Keep `jetpack/timeline-item` in the `widgets` category
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-e3d4a24841302670cbb36f6358268f27R15) `premium-content/button` from `layout` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-188c53da2d0b0b206b3ffca75cf1029bR15) `premium-content/login-button` from `layout` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-9988e0ed5ce04941b7958abf1bf5f8f5R18) `premium-content/container` from `common` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-f996a37875efc838ad070466ff8eeaf1R18) `premium-content/logged-out-view` from `common` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-15eafa59ca29a72c0c1539f1dc7c5666R14) `premium-content/subscriber-view` from `common` -> `design`

_*[0] This was a guess, let me know if this is the right category to move it to._
_*[1] I did set the category name for those using our helper function, even though I understand that a migration already took care of automatically mapping `layout` to `design` if I'm not mistaken. Let me know if you think it would not be needed for those cases._


#### TODO

- [X] Unit test for the helper function: the test itself is trivial, but I didn't add because I still need to learn more about how unit testing is done within calypso. I'll have a look at it soon.
- [X] Add support for TS for FSE tests.
- [X] Should we write tests to check that each block is assigned to the right category names depending on the set of categories? Not sure what is the testing policy for Calypso and what coverage we aim for. Any insights appreciated! (Jon said it's not needed)
- [x] Squash and rebase from master before merging

#### Outstanding questions

- Do we need to do anything with the category of the [old/deprecated](https://github.com/Automattic/wp-calypso/blob/master/apps/full-site-editing/full-site-editing-plugin/posts-list-block/blocks/posts-list/index.js#L36) `posts-list-block`? (More details in p1593038635493700-slack-cylon) 








#### Testing instructions

##### Atomated tests

* `cd` to `wp-calypso/apps/full-site-editing` and run `yarn test:js`.

##### To manually test the changes

1. Setup a simple FSE-enabled site and atomic ephemeral sites by following the instructions here: PCYsg-ly5-p2. Map it to your sandbox.
2. `cd` to `wp-calypso/apps/full-site-editing` and run `yarn dev --sync` to sync it to your sandbox.
3. Choose a theme that's compatible with FSE (i.e `Exford`)
4. Edit your home-page
5. Add a new block and search for the block you're testing. Verify it's in the right category. 

##### Test the scenarios above in the following environments:**
- [X] Simple website in calypso
- [X] Simple website in wp-admin
- [X] Atomic website
- [X] Atomic website with Gutenberg plugin disabled

~Note: If [this PR](https://github.com/WordPress/gutenberg/pull/23695) gets approved and merged, then I'll remove the helper function and tests from here (including the installation/configuration of jest).~
